### ISSUE TITLE
Deal with temporary network failures gracefully

### ISSUE THREAD

**Issue by mrmr1993** [CodaProtocol/coda#2879]
Last night I had an internet outage and my node was disconnected from all of the other peers. I restarted the daemon when my connection was back up and it failed with the following error:

```
    (monitor.ml.Error
  (runtime-lib/runtime.ml.E
    "Merkle roots differ after transfer: got unexpected result"
    ((expected
       35108718756611442719642669376718161310300000961643279755669423481426200000900832056256375702665615902882683721852963115403741412248152493965334437147424618522176921810559255690259361293761244570120814812510044267275086486720631)
      (got
        31905081918945275128429958096695184045061469120964658100951801106962483325159044009046028225749795960344499237893402942588382747468025201235871202511068745229274665783220516555487120951711036909619850504362619213488716602273659)
      (Loc lib/coda_base/ledger_transfer.ml:25:19)))
  ("Raised at file \"runtime-lib/runtime.ml\", line 28, characters 28-53"
    "Called from file \"runtime-lib/runtime.ml\" (inlined), line 105, characters 27-78"
    "Called from file \"lib/coda_base/ledger_transfer.ml\", line 25, characters 19-32"
    "Called from file \"lib/coda_lib/coda_lib.ml\", line 329, characters 4-73"
    "Called from file \"lib/coda_lib/coda_lib.ml\", line 371, characters 12-52"
    "Called from file \"src/deferred0.ml\", line 56, characters 64-69"
    "Called from file \"src/job_queue.ml\" (inlined), line 131, characters 2-5"
    "Called from file \"src/job_queue.ml\", line 170, characters 6-47"
    "Caught by monitor coda"))
```

Blowing away `~/.coda-config` obviously worked, but it might be better UX if we special-case this and try not to fail when we see that all peer connections have been lost.

**Comment by enolan** [comment]
I don't know what's going on with the Merkle roots, but when we can't
connect to a peer we punish it and remove it from our peer set. So in the
event of a network connection loss, if it's long enough you'll eventually
drop every peer from the in-memory peer set. Things will work again if you
restart the daemon but that's a bad user experience. Hmm, it's not obvious
what to do about this.

**Comment by mrmr1993** [comment]
Just noticed my node was down again, restarting gave the same style of error.

My node lost connection and removed all the peers from the set at around 14:00 UTC. Presumably the Merkle root change is because the state of the Merkle tree has moved on since then (it's now 17:40-ish UTC).

**Comment by enolan** [comment]
Solution for spuriously removing peers from the peer set: rather than removing them, we send them to "purgatory" and only try to contact nodes in purgatory if we can't connect to any live nodes. Nodes in purgatory get deprioritized if a connection attempt fails again.

**Comment by psteckler** [comment]
If we see a socket `ECONNREFUSED`, that would be grounds for moving a peer to this special set, right?

What about a Kademlia disconnect event?

**Comment by enolan** [comment]
Disconnects will happen if the peer happens to get evicted from one of the
k-buckets, which is normal if it's inactive. I think. Would need to check
the source/paper to be sure.

**Comment by psteckler** [comment]
Currently, the Kademlia disconnect event triggers a peer removal. If that reflects mere inactivity, I think we'd want to be able to try to connect to that peer again, if no others are available. So I'd say we should move that peer to the special set on disconnect.

Well, that's what I've implemented, so far.

**Comment by psteckler** [comment]
Fixed by #3005.
